pkg/proc/namedpipe_other.go

```diff
@@ -0,0 +1,84 @@
+//go:build !windows
+// +build !windows
+
+package proc
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+	"syscall"
+)
+
+type openOnRead struct {
+	p      string
+	rd     io.ReadCloser
+	status int32 // 0: initialization 1: file is opened 2: file is closed
+}
+
+func (oor *openOnRead) Read(p []byte) (n int, err error) {
+	if oor.rd != nil {
+		return oor.rd.Read(p)
+	}
+
+	// try from "into" to "open"
+	if !atomic.CompareAndSwapInt32(&oor.status, 0, 1) {
+		return 0, io.EOF
+	}
+
+	fh, err := os.OpenFile(oor.p, os.O_RDONLY, os.ModeNamedPipe)
+	if err != nil {
+		return 0, err
+	}
+
+	oor.rd = fh
+	return oor.rd.Read(p)
+}
+
+func (oor *openOnRead) Close() error {
+	defer os.Remove(oor.p)
+
+	// try from "init" to "close".
+	if !atomic.CompareAndSwapInt32(&oor.status, 0, 2) {
```

Comment:
This is awful, there has to be a better way. At minimum it should use a mutex instead.

Reply:
I think it should be enough to start the readers only after the target process has started.

Reply:
This can not work because rr.go (at minimum) would deadlock. However all we need to know is that the fifo has been opened for writing at least once, which we can do by unconditionally opening the fifo with `os.O_WRONLY|syscall.O_NONBLOCK`. This should work.

Reply:
> 
`os.O_WRONLY|syscall.O_NONBLOCK`  can solve the blocking problem. But, until the pipe is opened in write mode, the read pipe will return null.

https://go.dev/play/p/hKYuUNNg868

Reply:
I meant this:

```
func (oor *openOnRead) Close() error {
	defer os.Remove(oor.p)
	fh, _ := os.OpenFile(oor.p, os.O_WRONLY|syscall.O_NONBLOCK, 0)
	if fh != nil {
		fh.Close()
	}
	return oor.rd.Close()
}
```

Reply:
done.